tests/Integration/Behaviour/Features/Context/Domain/Product/UpdateProduct/UpdateOptionsFeatureContext.php

```diff
@@ -0,0 +1,155 @@
+    {
+        $this->updateProductOptions($productReference, $table, ShopConstraint::shop($this->getDefaultShopId()));
+    }
+
+    /**
+     * @When I update product :productReference options for shop ":shopReference" with following values:
+     *
+     * @param string $productReference
+     * @param string $shopReference
+     * @param TableNode $table
+     */
+    public function updateProductOptionsForShop(string $productReference, string $shopReference, TableNode $table): void
+    {
+        $shopId = $this->getSharedStorage()->get(trim($shopReference));
+        $this->updateProductOptions($productReference, $table, ShopConstraint::shop($shopId));
+    }
+
+    /**
+     * @When I update product :productReference options for all shops with following values:
+     *
+     * @param string $productReference
+     * @param TableNode $table
+     */
+    public function updateProductOptionsForAllShops(string $productReference, TableNode $table): void
+    {
+        $this->updateProductOptions($productReference, $table, ShopConstraint::allShops());
+    }
+
+    /**
+     * @When I assign non existing manufacturer to product :productReference
+     *
+     * @param string $productReference
+     */
+    public function updateOptionsWithNonExistingManufacturer(string $productReference): void
+    {
+        // intentional. Mimics id of non-existing manufacturer
+        $nonExistingId = 50000;
+
+        try {
+            $command = new UpdateProductCommand(
+                $this->getSharedStorage()->get($productReference),
+                ShopConstraint::shop($this->getDefaultShopId())
+            );
+            $command->setManufacturerId($nonExistingId);
+            $this->getCommandBus()->handle($command);
+        } catch (ManufacturerException $e) {
+            $this->setLastException($e);
+        }
+    }
+
+    /**
+     * @param string $productReference
+     * @param TableNode $table
+     * @param ShopConstraint $shopConstraint
+     */
+    protected function updateProductOptions(string $productReference, TableNode $table, ShopConstraint $shopConstraint): void
+    {
+        $data = $table->getRowsHash();
+        $productId = $this->getSharedStorage()->get($productReference);
+
+        try {
+            $command = new UpdateProductCommand($productId, $shopConstraint);
+            $this->fillCommand($data, $command);
+            $this->getCommandBus()->handle($command);
+        } catch (ProductException $e) {
+            $this->setLastException($e);
+        }
+    }
+
+    /**
+     * @param array $data
+     * @param UpdateProductCommand $command
+     */
+    private function fillCommand(array $data, UpdateProductCommand $command): void
+    {
+        if (isset($data['visibility'])) {
+            $command->setVisibility($data['visibility']);
+        }
+
+        if (isset($data['available_for_order'])) {
```

Comment:
When I see this code I'm thinking it's basically the same thing the CommandBuilder services are doing, maybe in a future refacto it would be worth using them here to build commands. It would validate both the handlers and the builder via behat

That's just an idea like this though, but it's worth discussing maybe